### ISSUE TITLE
Remove braces from switch

### ISSUE THREAD

**Issue by Pauan** [facebook/reason#772]
First off, I really appreciate this project. I've been vaguely interested in the ML-family for a long time, but the syntax was always a huge turnoff for me. Reason makes it a lot more palatable, to the point where I'm seriously considering using it for a big JavaScript project.

However, despite spending over 10 years programming in JavaScript, I find the `{}` in `switch` to be extremely ugly.

It's not a complete deal-breaker, but it's by far the biggest issue I have with Reason's syntax. Then again, I also like Haskell and Lisp syntax, so I'm definitely not a typical JavaScript developer! So take what I say with a grain of salt.

---

From what I understand, the reason for the braces is to properly disambiguate when you use nested `switch`

Couldn't you instead say that nested switches are a syntax error, and you are required to wrap the inner switch in parens?

In other words, this would be **invalid**:

```
switch foo
  | Bar =>
    switch bar
      | Qux =>
        ...
  | Corge =>
    ...
```

But this would be **valid**:

```
switch foo
  | Bar =>
    (switch bar
      | Qux =>
        ...)
  | Corge =>
    ...
```

Would that solve the issue? Or am I missing something?

**Comment by yunxing** [comment]
Thanks for the input, @Pauan. Always love to see feedbacks and ideas. 

Your concern about nested switch statement is right and the proposed solution would work. 

Note that in this case, we will need to add parens to all expression that could contain a "|", including cases like:
```
switch foo
  | Bar =>
    (try bar
      | Qux =>
        ...)
```

```
switch foo
  | Bar =>
    (fun
      | Qux =>
        ...)
```

```
switch foo
  | Bar =>
    (if true 1 else switch
      | Qux =>
        ...)
```
and since the parse can only look ahead for 1 token, it means even if an expression don't contain "|", as long as it *could potentially* contain "|", we need to add a pair of parens to it:
```
switch foo
  | Bar =>
    (if true 1 else 2)
 | Corge =>
```

to resolve this issue where parens are needed even when "|" is not present, we can make your "invalid" example somehow "valid" by *parsing* them by default as 

```
switch foo
  | Bar =>
    (switch bar
      | Qux =>
        ...
  | Corge =>
    ...
)
```

This is what the vanilla ocaml parser does, as far as I can remember.

**Comment by Pauan** [comment]
Ah, I hadn't considered a `switch` nested inside of an `if`

You said that the parser can only lookahead one token, but would it be possible for the parser to keep track of previous context?

In other words, it would remember that "I am parsing inside of a switch", so when it encounters another switch (no matter how deeply nested), it will only require parens around the inner `switch`?

In other words, you would need to write code like this:

```
switch foo
  | Bar =>
    if true 1 else (switch
      | Qux =>
        ...)
```

However, if there isn't any ambiguity (e.g. because the inner `switch` is inside of a block), then it wouldn't require the parens, therefore this would be valid:

```
switch foo
  | Bar =>
    if true {
      1
    } else {
      switch
        | Qux =>
          ...
    }
```

If I'm understanding correctly, this doesn't require arbitrary lookahead, but it does require keeping track of state (i.e. whether we're within a `switch` and not within a block).

**Comment by yunxing** [comment]
@Pauan You are totally right. The parser indeed keeps states on a stack. I can imagine if we go down this route, we can change the our implementation of parser (which is in reason_parser.mly) so that instead of looking for a general "expression" object when parsing "if statement", look for an "expression_for_if" object where `switch` requires parens. We have used the same design pattern in other parts of our implementation as well: Take a look at "simple_expression" which requires a parens if the expression may contain any space. This pattern does make the implementation a bit harder to understand, so we should be very careful in introducing it.

**Comment by mrkaspa** [comment]
Would love to see this also for if, almost all the modern languages have removed the parentheses from the if like rust, go, elixir

**Comment by jaredly** [comment]
gonna close this for now -- I'm pretty sure we won't remove the braces from switch
